File: 08/problem.py
def getInput(file):
    data = None

    with open(file) as f:
        data = f.read().strip().split('\n')

    return data

def printMap(treeMap):
    idx = 0
    for i in treeMap:
        for j in i:
            print(j, end=" ")
        print()

def calcScenicScore(row, col, treeMap):
    rightScore = 0
    leftScore  = 0
    upScore = 0
    downScore = 0

    for i in range(col+1, len(treeMap[row])):
        if (treeMap[row][col] > treeMap[row][i]):
            rightScore += 1
        else:
            rightScore += 1
            break
    
    # look left
    for i in range(col-1, -1, -1):
        if (treeMap[row][col] > treeMap[row][i]):
            leftScore += 1
        else:
            leftScore += 1
            break
    # look up

    for i in range(row-1, -1, -1):
        if (treeMap[row][col] > treeMap[i][col]):
            upScore += 1
        else:
            upScore += 1
            break

    for i in range(row+1, len(treeMap)):
        if (treeMap[row][col] > treeMap[i][col]):
            downScore += 1
        else:
            downScore += 1
            break
    
    return rightScore * leftScore * upScore * downScore

def isVisible(row, col, treeMap):
    visibleRight = True
    visibleLeft = True
    visibleUp = True
    visibleDown = True

    # look right
    for i in range(col+1, len(treeMap[row])):
        if (treeMap[row][col] <= treeMap[row][i]):
            visibleRight = False
    # look left
    for i in range(col-1, -1, -1):
        if (treeMap[row][col] <= treeMap[row][i]):
            visibleLeft = False
    # look up
    for i in range(row-1, -1, -1):
        if (treeMap[row][col] <= treeMap[i][col]):
            visibleUp = False
    # look right
    for i in range(row+1, len(treeMap[row])):
        if (treeMap[row][col] <= treeMap[i][col]):
            visibleDown = False

    return visibleUp or visibleDown or visibleLeft or visibleRight

def solveFirst(data):
    treeMap = []
    visible = 0
    for line in data:
        row = []
        for tree in line:
            row.append(tree)
        treeMap.append(row)
    
    alwaysVisI = [0, len(treeMap[0])-1]
    alwaysVisJ = [0, len(treeMap[0])-1]
    for i in range(len(treeMap)):
        for j in range(len(treeMap[i])):
            if (i in alwaysVisI or j in alwaysVisJ):
                visible += 1
            else:
                if(isVisible(i, j, treeMap)):
                    visible += 1
    print(visible)

def solveSecond(data):
    treeMap = []
    visible = 0
    for line in data:
        row = []
        for tree in line:
            row.append(tree)
        treeMap.append(row)
    
    alwaysVisI = [0, len(treeMap[0])-1]
    alwaysVisJ = [0, len(treeMap[0])-1]
    maxScore = 0
    for i in range(len(treeMap)):
        for j in range(len(treeMap[i])):
            res = calcScenicScore(i, j, treeMap)
            if (res > maxScore):
                maxScore = res
    
    print(maxScore)

if __name__ == "__main__":
    data = getInput("./input")
    solveFirst(data)
    solveSecond(data)

File: 08/problem.lua
local function split (inputstr, sep)
    if sep == nil then
       sep = "%s"
    end
    local t={}
    for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
       table.insert(t, str)
    end
    return t
end

local function getInput(fileName)
    local fileIn = io.open(fileName)

    -- check if file was read successfully
    if (fileIn == nil)
    then
        return
    end

    -- this has state, so can't use twice for each problem 
    local lines = fileIn:lines()
    local count = 1
    local returnLines = {}

    -- i want a stateless iterator, theres probably a better way to do this but it works
    for line in lines do
        returnLines[count] = line
        count = count + 1
    end

    return returnLines
end

function PrintTable(table)

    for idx, value in pairs(table) do
        for idx2, value2 in pairs(value) do
            io.write(value2, " ")
        end
        io.write("\n")
    end

end
-- i have to write a set function??? lul
function Set(str)
    local s = {}
    for char in str:gmatch"." do s[char] = true end
    return s
end

function IsUpper(char)
    return string.match(char, '%u')
end

local function isVisible(row, col, treeMap)

    print("")

end

local function solveFirst(input)
    -- do things here to solve problem
    local treeMap = {}
    for idx, line in pairs(input) do
        local row = {}
        for char in line:gmatch"." do
            table.insert(row, char)
        end
        table.insert(treeMap, row)
    end
    
    PrintTable(treeMap)
    local visibleScore = 0
    for idx, value in pairs(treeMap) do
        for idx2, value2 in pairs(treeMap[1]) do
            if (idx == 0 or idx2 == 0 or idx == #treeMap or idx2 == #treeMap[1]) then
                visibleScore = visibleScore + 1
            else 
                
            end
        end
    end

    print(visibleScore)
    return visibleScore
    -- do things above here to solve the problem
end

local function solveSecond(input)
    -- do things here to solve problem
    for idx, line in pairs(input) do
        print(line)
    end
    -- do things above here to solve the problem
end

local input = getInput("./test-input")
solveFirst(input)
-- solveSecond(input)
